airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py

```diff
@@ -0,0 +1,463 @@
+
+    entity_prefix = "ad"
+    enable_deleted = True
+
+    @backoff_policy
+    def _read_records(self, params: Mapping[str, Any]):
+        return self._api.account.get_ads(params=params, fields=[self.cursor_field])
+
+
+class AdSets(FBMarketingIncrementalStream):
+    """doc: https://developers.facebook.com/docs/marketing-api/reference/ad-campaign"""
+
+    entity_prefix = "adset"
+    enable_deleted = True
+
+    @backoff_policy
+    def _read_records(self, params: Mapping[str, Any]):
+        return self._api.account.get_ad_sets(params=params)
+
+
+class Campaigns(FBMarketingIncrementalStream):
+    """doc: https://developers.facebook.com/docs/marketing-api/reference/ad-campaign-group"""
+
+    entity_prefix = "campaign"
+    enable_deleted = True
+
+    @backoff_policy
+    def _read_records(self, params: Mapping[str, Any]):
+        return self._api.account.get_campaigns(params=params)
+
+
+class AdsInsights(FBMarketingIncrementalStream):
+    """doc: https://developers.facebook.com/docs/marketing-api/insights"""
+
+    primary_key = None
+
+    ALL_ACTION_ATTRIBUTION_WINDOWS = [
+        "1d_click",
+        "7d_click",
+        "28d_click",
+        "1d_view",
+        "7d_view",
+        "28d_view",
+    ]
+
+    ALL_ACTION_BREAKDOWNS = [
+        "action_type",
+        "action_target_id",
+        "action_destination",
+    ]
+
+    MAX_WAIT_TO_START = pendulum.duration(minutes=5)
+    MAX_WAIT_TO_FINISH = pendulum.duration(minutes=30)
+    MAX_ASYNC_SLEEP = pendulum.duration(minutes=5)
+    MAX_ASYNC_JOBS = 30
+    INSIGHTS_RETENTION_PERIOD = pendulum.duration(days=37 * 30)
+
+    action_breakdowns = ALL_ACTION_BREAKDOWNS
+    level = "ad"
+    action_attribution_windows = ALL_ACTION_ATTRIBUTION_WINDOWS
+    time_increment = 1
+
+    breakdowns = []
+
+    def __init__(self, buffer_days, days_per_job, **kwargs):
+        super().__init__(**kwargs)
+        self.lookback_window = pendulum.Duration(days=buffer_days)
+        self._days_per_job = days_per_job
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        """Waits for current job to finish (slice) and yield its result"""
+        result = self.wait_for_job(stream_slice["job"])
+        # because we query `lookback_window` days before actual cursor we might get records older then cursor
+        stream_state = stream_state or {}
+        state_value = stream_state.get(self.cursor_field)
+        min_cursor = self._start_date if not state_value else pendulum.parse(state_value)
+
+        for obj in result.get_result():
+            if pendulum.parse(obj[self.cursor_field]) >= min_cursor:
+                yield obj.export_all_data()
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
```

Comment:
is retry functionality working correctly? We never retry a job if it fails. is that the desired behavior?

Reply:
usually, it is a very good reason for a job to fail. At least the response doesn't contain any useful information about the reason, in my practice failed job appears only when something is wrong with the query itself (fields or breakdowns not supported), so here we only retry checking of the status of the job. We save states for every async job, so if the fails we have a checkpoint.

Reply:
https://developers.facebook.com/docs/marketing-api/insights/best-practices/
from the docs
```
Job Failed | Job has failed. Review your query and try again.
```

Reply:
Facebook is leading by counter-example on how to give a good error message